### PR TITLE
Rewrite: 1.0.0-alpha, with full statement support, fluid syntax

### PR DESCRIPTION
**Pull request by matanlurey**
This is a large refactor of of `code_builder` to able to replace the Angular Output AST code.

Still missing before PR ready for review:
- [x] Re-add scoping support
- [x] More e2e-style tests that existed in v0 branch

/cc @yjbanov @hterkelsen

### CODE REVIEW

File: lib/src/builders/annotation.dart [hunk 1]
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/analyzer.dart';
+import 'package:code_builder/src/builders/class.dart';
+import 'package:code_builder/src/builders/parameter.dart';
+import 'package:code_builder/src/builders/shared.dart';
+
+/// Lazily builds an [Annotation] AST when [buildAnnotation] is invoked.
```
Comment (yjbanov) [inline lib/src/builders/annotation.dart]: I'd remove "Lazily" from the comments throughout the package. This word only makes me wonder what it means and might scare me away from using this class :)

File: lib/src/builders/parameter.dart [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+import 'package:analyzer/analyzer.dart';
+import 'package:code_builder/code_builder.dart';
+import 'package:code_builder/src/builders/annotation.dart';
+import 'package:code_builder/src/builders/method.dart';
+import 'package:code_builder/src/builders/shared.dart';
+import 'package:code_builder/src/builders/type.dart';
+import 'package:code_builder/src/tokens.dart';
+
+/// A more short-hand way of constructing a [ParameterBuilder].
```
Comment (hterkelsen) [inline lib/src/builders/parameter.dart]: get rid of "more", it's cleaner

File: lib/src/builders/type.dart [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    Iterable<ExpressionBuilder> positional, [
+    Map<String, ExpressionBuilder> named = const {},
+  ]) {
+    final builder = new NewInstanceBuilder._const(this);
+    _addArguments(builder, positional, named);
+    return builder;
+  }
+
+  /// Invokes `const` on this type with a [name]d constructor.
+  NewInstanceBuilder constInstanceWith(
```
Comment (hterkelsen) [inline lib/src/builders/type.dart]: constInstanceWith -> namedConstInstance?

File: lib/src/scope.dart [hunk 4]
```diff
@@ -2,81 +2,115 @@
...
 ///     void useContext(Scope scope) {
-///       // Prints Identifier "i1.Foo"
-///       print(scope.getIdentifier('Foo', 'package:foo/foo.dart');
+///       // May print 'i1.Foo'.
+///       print(scope.identifier('Foo', 'package:foo/foo.dart'));
 ///
-///       // Prints Identifier "i1.Bar"
-///       print(scope.getIdentifier('Bar', 'package:foo/foo.dart');
+///       // May print 'i1.Bar'.
+///       print(scope.identifier('Bar'), 'package:foo/foo.dart');
```
Comment (hterkelsen) [inline lib/src/scope.dart]: here and in the `Baz` example, the closing parenthesis should be after the package, not the identifier

File: lib/src/scope.dart [hunk 5]
```diff
@@ -2,81 +2,115 @@
...
+  /// **WARNING**: Does not collect import statements. This is only really
+  /// advisable for use in tests but not production code. To use import
+  /// collection but not prefixing, see [Scope.dedupe].
+  static const Scope identity = const _IdentityScope();
 
-  /// Create a context that just de-duplicates imports (no scoping).
-  factory Scope.dedupe() = _DeduplicatingScope;
+  /// Create a new scoping context.
+  ///
+  /// Actual implementation of is _not_ guaranteed, only that all import
```
Comment (hterkelsen) [inline lib/src/scope.dart]: Actual implementation of what?

File: .gitignore [hunk 6]
```diff
@@ -8,23 +6,9 @@
...
-*.info.json
-
 # Directory created by dartdoc
 doc/api/
 
 # Don't commit pubspec lock file
 # (Library packages only! Remove pattern if developing an application package)
 pubspec.lock
-
-*.iml
```
Comment (yjbanov) [inline .gitignore]: There's this issue to watch out for: https://github.com/flutter/flutter-intellij/issues/302

File: README.md [hunk 7]
```diff
@@ -34,10 +35,14 @@ Code builder has a narrow and user-friendly API.
...
-new ClassBuilder('Animal', extends: 'Organism')
-  ..addMethod(new MethodBuilder.returnVoid('eat')
-    ..setExpression(new ExpressionBuilder.invoke('print',
-      positional: [new LiteralString('Yum!')])));
+var base = reference('Organism');
+var clazz = new ClassBuilder('Animal', asExtends: base);
+clazz.addMethod(
+  new MethodBuilder.returnVoid(
+    'eat',
+    returns: reference('print).call([literal('Yum')]),
```
Comment (matanlurey) [inline README.md]: done

File: lib/dart/async.dart [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+///
+/// First import the library:
+///     import 'package:code_builder/code_builder.dart';
+///     import 'package:code_builder/dart/async.dart';
+///
+/// All references are _namespaced_ under [async]. Try it:
+///     // Outputs: new Future.value('Hello')
+///     async.Future.newInstanceNamed('value', [literal('Hello')]);
+///
+/// If you are [missing a reference from `dart:async`](https://goo.gl/XbSfmT)
```
Comment (matanlurey) [inline lib/dart/async.dart]: done

File: lib/dart/async.dart [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+/// please send us a [pull request](https://goo.gl/2LvV7f) or
+/// [file an issue](https://goo.gl/IooPfl).
+library code_builder.dart.async;
+
+import 'dart:async' as dart_async;
+
+import 'package:code_builder/code_builder.dart';
+
+/// A namespace for references in `dart:async`.
+final DartAsync async = new DartAsync._();
```
Comment (matanlurey) [inline lib/dart/async.dart]: done

File: lib/dart/async.dart [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+/// If you are [missing a reference from `dart:async`](https://goo.gl/XbSfmT)
+/// please send us a [pull request](https://goo.gl/2LvV7f) or
+/// [file an issue](https://goo.gl/IooPfl).
+library code_builder.dart.async;
+
+import 'dart:async' as dart_async;
+
+import 'package:code_builder/code_builder.dart';
+
+/// A namespace for references in `dart:async`.
```
Comment (matanlurey) [inline lib/dart/async.dart]: done

File: lib/src/builders/annotation.dart [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  /// Returns a [List] of all built [Annotation]s.
+  List<Annotation> buildAnnotations([Scope scope]) => _annotations
+      .map/*<Annotation>*/((a) => a.buildAnnotation(scope))
+      .toList();
+
+  /// Clones all annotations to [clone].
+  void cloneAnnotationsTo(HasAnnotations clone) {
+    clone.addAnnotations(_annotations);
```
Comment (matanlurey) [inline lib/src/builders/annotation.dart]: Renamed as "copyTo"

File: lib/src/builders/class.dart [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+      clazz.addMethod(member);
+    } else {
+      throw new StateError('Invalid AST type: ${member.runtimeType}');
+    }
+  }
+  return clazz;
+}
+
+/// Wrap [member] to be emitted as a `static` method or field.
+_StaticFieldWrapper asStatic(ValidClassMember member) {
```
Comment (matanlurey) [inline lib/src/builders/class.dart]: done (former)

File: lib/src/builders/shared.dart [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+/// Returns a string [Literal] from [value].
+Identifier stringIdentifier(String value) =>
+    new SimpleIdentifier(stringToken(value));
+
+/// Lazily builds an analyzer [AstNode] when [buildAst] is invoked.
+///
+/// Most builders should also have specific typed methods for returning their
+/// type of AST node, such as `buildExpression` for returning an `Expression`
+/// AST.
+abstract class AstBuilder<T extends AstNode> {
```
Comment (matanlurey) [inline lib/src/builders/shared.dart]: Not too crazy. Maybe too crazy for 1.0 though :)

File: .gitignore (hunk 6 above)
Comment (matanlurey) [inline .gitignore]: I was advised to just use a global gitignore for my IDEs

File: lib/code_builder.dart [hunk 14]
```diff
@@ -2,69 +2,47 @@
...
+export 'src/builders/parameter.dart' show parameter, ParameterBuilder;
+export 'src/pretty_printer.dart' show prettyToSource;
+export 'src/builders/reference.dart'
+    show explicitThis, reference, ReferenceBuilder;
+export 'src/builders/shared.dart' show AstBuilder, Scope;
+export 'src/builders/statement.dart'
+    show ifThen, elseIf, elseThen, IfStatementBuilder, StatementBuilder;
+export 'src/builders/type.dart' show NewInstanceBuilder, TypeBuilder;
+
+final _dartFmt = new DartFormatter();
```
Comment (matanlurey) [inline lib/code_builder.dart]: done

File: lib/dart/async.dart [hunk 15]
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// Contains reference to the `dart:async` SDK for use in code generation.
```
Comment (matanlurey) [inline lib/dart/async.dart]: done

File: lib/src/builders/expression.dart [hunk 16]
```diff
@@ -0,0 +1,349 @@
...
+        return new MapLiteralEntry(_literal(k), $colon, _literal(value[k]));
+      }).toList(),
+      $closeBracket,
+    );
+  }
+  throw new ArgumentError.value(value, 'Unsupported');
+}
+
+/// Implements much of [ExpressionBuilder].
+abstract class AbstractExpressionMixin implements ExpressionBuilder {
```
Comment (matanlurey) [inline lib/src/builders/expression.dart]: Done

File: lib/src/builders/expression.dart [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+    Map<String, ExpressionBuilder> namedArguments = const {},
+  ]) {
+    final invocation = new InvocationBuilder._on(this, method);
+    positionalArguments.forEach(invocation.addPositionalArgument);
+    namedArguments.forEach(invocation.addNamedArgument);
+    return invocation;
+  }
+
+  @override
+  ExpressionBuilder negate() => new _NegateExpression(this);
```
Comment (matanlurey) [inline lib/src/builders/expression.dart]: Done

File: lib/src/builders/field.dart [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+  ExpressionBuilder value,
+}) {
+  return new FieldBuilder(
+    name,
+    type: type,
+    value: value,
+  );
+}
+
+/// Short-hand for [FieldBuilder].
```
Comment (matanlurey) [inline lib/src/builders/field.dart]: Done

File: lib/src/builders/parameter.dart (hunk 2 above)
Comment (matanlurey) [inline lib/src/builders/parameter.dart]: Done

File: lib/src/builders/type.dart (hunk 3 above)
Comment (matanlurey) [inline lib/src/builders/type.dart]: Done

File: lib/src/scope.dart (hunk 4 above)
Comment (matanlurey) [inline lib/src/scope.dart]: Done

File: lib/src/scope.dart (hunk 5 above)
Comment (matanlurey) [inline lib/src/scope.dart]: Clarified

File: lib/dart/async.dart (hunk 8 above)
Comment (matanlurey) [inline lib/dart/async.dart]: Done

File: lib/dart/async.dart (hunk 10 above)
Comment (matanlurey) [inline lib/dart/async.dart]: Done

File: lib/dart/async.dart (hunk 9 above)
Comment (matanlurey) [inline lib/dart/async.dart]: Done

File: lib/src/builders/annotation.dart (hunk 1 above)
Comment (matanlurey) [inline lib/src/builders/annotation.dart]: Done

File: lib/src/builders/class.dart (hunk 12 above)
Comment (matanlurey) [inline lib/src/builders/class.dart]: Done

**Comment by matanlurey** [comment]
@googlebot: This was me :-(

**Pull request merged**